internal/util/cgo/futures.go

```diff
@@ -0,0 +1,182 @@
+	future_register_ready_callback(f, unlockMutexOnC, m);
+}
+*/
+import "C"
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"unsafe"
+
+	"github.com/cockroachdb/errors"
+
+	"github.com/milvus-io/milvus/pkg/util/merr"
+)
+
+var ErrConsumed = errors.New("future is already consumed")
+
+// Would put this in futures.go but for the documented issue with
+// exports and functions in preamble
+// (https://code.google.com/p/go-wiki/wiki/cgo#Global_functions)
+//
+//export unlockMutex
+func unlockMutex(p unsafe.Pointer) {
+	m := (*sync.Mutex)(p)
+	m.Unlock()
+}
+
+type basicFuture interface {
+	// Context return the context of the future.
+	Context() context.Context
+
+	// BlockUntilReady block until the future is ready or canceled.
+	// caller can call this method multiple times in different concurrent unit.
+	BlockUntilReady()
+
+	// cancel the future with error.
+	cancel(error)
+}
+
+type Future interface {
+	basicFuture
+
+	// BlockAndLeakyGet block until the future is ready or canceled, and return the leaky result.
+	//   Caller should only call once for BlockAndLeakyGet, otherwise the ErrConsumed will returned.
+	//   Caller will get the merr.ErrSegcoreCancel or merr.ErrSegcoreTimeout respectively if the future is canceled or timeout.
+	//   Caller will get other error if the underlying cgo function throws, otherwise caller will get result.
+	//   Caller should free the result after used (defined by caller), otherwise the memory of result is leaked.
+	BlockAndLeakyGet() (unsafe.Pointer, error)
+}
+
+type (
+	CFuturePtr       unsafe.Pointer
+	CGOAsyncFunction = func() CFuturePtr
+)
+
+// Async is a helper function to call a C async function that returns a future.
+func Async(ctx context.Context, f CGOAsyncFunction, opts ...Opt) Future {
+	options := getDefaultOpt()
+	// apply options.
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	// create a future for caller to use.
+	var cFuturePtr *C.CFuture
+	getCGOCaller().call(options.name, func() {
+		cFuturePtr = (*C.CFuture)(f())
+	})
+
+	ctx, cancel := context.WithCancel(ctx)
+	future := &futureImpl{
+		closure:      f,
+		ctx:          ctx,
+		ctxCancel:    cancel,
+		releaserOnce: sync.Once{},
+		future:       cFuturePtr,
+		opts:         options,
+		state:        newFutureState(),
+	}
+
+	runtime.SetFinalizer(future, func(future *futureImpl) {
+		getCGOCaller().call("future_destroy", func() {
+			C.future_destroy(future.future)
+		})
+	})
+
+	// register the future to do timeout notification.
+	futureManager.Register(future)
+	return future
+}
+
+type futureImpl struct {
+	ctx          context.Context
+	ctxCancel    context.CancelFunc
+	future       *C.CFuture
+	closure      CGOAsyncFunction
+	opts         *options
+	state        futureState
+	releaserOnce sync.Once
+}
+
+func (f *futureImpl) Context() context.Context {
+	return f.ctx
+}
+
+func (f *futureImpl) BlockUntilReady() {
+	f.blockUntilReady()
+}
+
+func (f *futureImpl) BlockAndLeakyGet() (unsafe.Pointer, error) {
+	f.blockUntilReady()
```

Comment:
same here, blockUntil ready and future leak get and get and be one cgo?

Reply:
- blockUntilready can be called multiple times, but leakyget can be only call once. It's better to implement two method to perform a future-like api.
- we cannot merge it together, there are always two cgo call will be applied `future_go_register_ready_callback`, ``future_leak_and_get` to avoid long cgo call.